packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts

```diff
@@ -59,42 +60,53 @@ async function createDataObject<TObj extends PureDataObject, I extends DataObjec
+            }
+            return createResponseError(500, "Data store runtime handle is not an IFluidRouter", request);
+        },
+        runtimeClass);
 
     // Create a new runtime for our data store
     // The runtime is what Fluid uses to create DDS' and route to your data store
-    const runtime = new runtimeClass(
+    const runtime: FluidDataStoreRuntime = new runtimeClass(
         context,
         sharedObjectRegistry,
         existing,
-    );
+        async (rt: IFluidDataStoreRuntime) => {
+            // Create the data object as part of initializing the handle for the data store runtime. This allows it
+            // to register various callbacks with the runtime before the runtime becomes globally available. However,
+            // this is not full initialization - the constructor can't access DDSes or other services of runtime as
+            // objects are not fully initialized. In order to use object, we need to go through full initialization
+            // by calling finishInitialization().
+            const scope: FluidObject<IFluidDependencySynthesizer> = context.scope;
+            const dependencyContainer = new DependencyContainer(scope.IFluidDependencySynthesizer);
+            const providers = dependencyContainer.synthesize<I["OptionalProviders"]>(optionalProviders, {});
+            const instance = new ctor({ runtime: rt, context, providers, initProps });
 
-    // Create object right away.
-    // This allows object to register various callbacks with runtime before runtime
-    // becomes globally available. But it's not full initialization - constructor can't
-    // access DDSes or other services of runtime as objects are not fully initialized.
-    // In order to use object, we need to go through full initialization by calling finishInitialization().
-    const scope: FluidObject<IFluidDependencySynthesizer> = context.scope;
-    const dependencyContainer = new DependencyContainer(scope.IFluidDependencySynthesizer);
-    const providers = dependencyContainer.synthesize<I["OptionalProviders"]>(optionalProviders, {});
-    const instance = new ctor({ runtime, context, providers, initProps });
+            // If it's a newly created object we need to wait for it to finish initialization (as that results in
+            // creation of DDSes) before it gets attached, providing atomic guarantee of creation.
+            // WARNING: we can't do the same (yet) for already existing PureDataObject!
+            // This will result in deadlock, as it tries to resolve internal handles, but any
+            // handle resolution goes through root (container runtime), which can't route it back
+            // to this data store, as it's still not initialized and not known to container runtime yet.
+            // In the future, we should address it by using relative paths for handles and be able to resolve
+            // local DDSes while data store is not fully initialized.
+            if (!existing) {
```

Comment:
I think the point of checking `existing` is moot here. If we already requesting DataObject, then it needs to be initialized (on all code paths). I'm sure old request flow was doing it all the time.
At the same time moving initialization here makes it delayed, and I think it will break scenario where data store is created and aliased (i.e. it becomes root data store). I think that's what changes in datastoreRuntime.ts attempt to do, but I think it's a wrong place to do so, I'd rather see DataObject own this (as it needs it). Maybe the easiest way to accomplish it is ensured that createRootInstance() flows ensure DataObject is fully initialized (though I think it leaves a gap of create DataObject that is not root, and then using aliasing API from runtime directly)

Reply:
i think we can make this change much less invasive by basically keeping the code as is, but making instance a non-constant the simply gets returned by the entrypoint function, and then gets it value set after the runtime is created, to be extra safe you can also assert in the entrypoint function that instance isn't undefined. i'd really like to minimize churn at this point in time. this discussion does point to concerns we will want to address, but we shouldn't do everything, even if good things, at once.

Reply:
that will also allow instance to be available in the functions that call this function, which will even further reduce churn